lib/stm32/h7/rcc.c

```diff
@@ -6,16 +6,333 @@
+	rcc_clock_tree.per.pclk2_mhz = rcc_prediv_3bit_log_div(rcc_clock_tree.hclk_mhz,
+		config->domain2.pclk2_prescale >> RCC_D2CFGR_D2PPRE2_SHIFT);
+	rcc_clock_tree.per.pclk1_mhz = rcc_prediv_3bit_log_div(rcc_clock_tree.hclk_mhz,
+		config->domain2.pclk1_prescale >> RCC_D2CFGR_D2PPRE1_SHIFT);
+}
+
+static void rcc_clock_setup_domain3(const struct rcc_pll_config *config) {
+	RCC_D3CFGR &= RCC_D3CFGR_RSVD_BITMASK;
+	RCC_D3CFGR |= config->domain3.pclk4_prescale;
+
+	/* Update our clock values in our tree based on the config values. */
+	rcc_clock_tree.per.pclk4_mhz = rcc_prediv_3bit_log_div(rcc_clock_tree.hclk_mhz,
+		config->domain3.pclk4_prescale >> RCC_D3CFGR_D3PPRE_SHIFT);
+}
+
+void rcc_clock_setup_pll(const struct rcc_pll_config *config) {
+	/* First, set system clock to utilize HSI, then disable all but HSI. */
+	RCC_CR |= RCC_CR_HSION;
+	RCC_CFGR &= ~(RCC_CFGR_SW_MASK << RCC_CFGR_SW_SHIFT);
+	while (((RCC_CFGR >> RCC_CFGR_SWS_SHIFT) & RCC_CFGR_SWS_MASK) != RCC_CFGR_SWS_HSI);
+	RCC_CR = RCC_CR_HSION;
+
+	/* User has specified an external oscillator, make sure we turn it on. */
+	if (config->hse_frequency > 0) {
+		RCC_CR |= RCC_CR_HSEON;
+		while (!(RCC_CR & RCC_CR_HSERDY));
+		rcc_clock_tree.hse_khz = config->hse_frequency / HZ_PER_KHZ;
+	}
+
+	/* Set, enable and lock all of the pll from the config. */
+	rcc_set_and_enable_plls(config);
+
+	/* Populate our base sysclk settings for use with domain clocks. */
+	if (config->sysclk_mux == RCC_SYSCLK_PLL) {
+		rcc_clock_tree.sysclk_mhz = rcc_clock_tree.pll[PLL1_INDEX].p_mhz;
+	} else if (config->sysclk_mux == RCC_SYSCLK_HSE) {
+		rcc_clock_tree.sysclk_mhz = config->hse_frequency / HZ_PER_MHZ;
+	} else {
+		rcc_clock_tree.sysclk_mhz = HSI_BASE_FREQUENCY / HZ_PER_MHZ;
+	}
+
+	/* PLL's are set, now we need to get everything switched over the correct domains. */
+	rcc_clock_setup_domain1(config);
+	rcc_clock_setup_domain2(config);
+	rcc_clock_setup_domain3(config);
+
+	/* TODO: Configure custom kernel mappings. */
+
+	/* Domains dividers are all configured, now we can switchover to PLL. */
+	RCC_CFGR |= RCC_CFGR_SW_PLL1;
+	uint32_t cfgr_sws = ((RCC_CFGR >> RCC_CFGR_SWS_SHIFT) & RCC_CFGR_SWS_MASK);
+	while(cfgr_sws != RCC_CFGR_SWS_PLL1) {
+		cfgr_sws = ((RCC_CFGR >> RCC_CFGR_SWS_SHIFT) & RCC_CFGR_SWS_MASK);
+	}
+}
+
+uint32_t rcc_get_clock(enum rcc_clock_source source) {
+	uint32_t clksel;
+	switch (source) {
+		case RCC_SYSCLK:
+			return rcc_clock_tree.sysclk_mhz * HZ_PER_MHZ;
+		case RCC_CPUCLK:
+		case RCC_SYSTICKCLK:
+			return rcc_clock_tree.cpu_mhz * HZ_PER_MHZ;
+		case RCC_AHBCLK:
+		case RCC_HCLK3:
+			return rcc_clock_tree.hclk_mhz * HZ_PER_MHZ;
+		case RCC_PCLK1:
+		case RCC_APB1CLK:
+			return rcc_clock_tree.per.pclk1_mhz * HZ_PER_MHZ;
+		case RCC_PCLK2:
+		case RCC_APB2CLK:
+			return rcc_clock_tree.per.pclk2_mhz * HZ_PER_MHZ;
+		case RCC_PCLK3:
+		case RCC_APB3CLK:
+			return rcc_clock_tree.per.pclk3_mhz * HZ_PER_MHZ;
+		case RCC_PCLK4:
+		case RCC_APB4CLK:
```

Comment:
can't say I'm a fan of having duplicate names here.  Not sure what it gets us other than confusion and inconsistency if some places only take one or the other.

Reply:
This is mainly because the documentation refers to them with both names. The idea is that if the user is digging through the docs, and it says one or the other, they don't have to dig through the tree documentation to discover that the docs refer to PCLK4 (or per_clk4) which is the same thing as APB4, or vice versa.  It was unclear which of these will be the "correct" name to use if I were to eliminate one of them. Happy to eliminate one, but would like opinions on which to use.

Reply:
I'd go with APB, as that's the name used in other parts, both ST and other cortex.  At some point people have to know what they're doing, and I think allowing/encouraing mix and match here makes things worse not better.